errors.go

```diff
@@ -0,0 +1,164 @@
+package jsonapi
+
+import (
+	"encoding/json"
+	"fmt"
+	"io"
+)
+
+// MarshalErrors writes a JSON API response using the given `[]error`.
+// This function works by passing each `error` in the given slice through a routine which attempts
+// to cast the error to each of this package's `Error<FieldName>Compatible` interfaces. Though
+// it works for errors which do not implement any of these interfaces as well. For more information
+// on JSON API error payloads, see the spec here: http://jsonapi.org/format/#document-top-level
+// and here: http://jsonapi.org/format/#error-objects
+//
+// To use this function in your code effectively, make your package's error types implement some
+// or all of the error compatibility interfaces (`ErrorTitleCompatible`, `ErrorDetailCompatible`,
+// ...). Doing so gives you direct control over what will appear in the serialied error object's
+// fields. Without implementing these interfaces, the `title` & `detail` fields will be populated
+// with derived data about the error.
+func MarshalErrors(w io.Writer, errs []error) error {
+	// Serialize the given errors.
+	var formattedErrors []ErrorObject
+	for _, err := range errs {
+		e := marshalError(err)
+		formattedErrors = append(formattedErrors, e)
+	}
+
+	// Write out the serialize errors payload.
+	if err := json.NewEncoder(w).Encode(&ErrorsPayload{Errors: formattedErrors}); err != nil {
+		return err
+	}
+	return nil
+}
+
+// marshalError will serialize the given error as best as possible according to this
+// package's `Error<field>Compatible` interfaces.
+func marshalError(err error) ErrorObject {
+	errorObject := ErrorObject{}
+	if e, ok := err.(ErrorIDCompatible); ok {
+		errorObject.ID = e.GetID()
+	}
+
+	if e, ok := err.(ErrorTitleCompatible); ok {
+		errorObject.Title = e.GetTitle()
+	} else {
+		errorObject.Title = fmt.Sprintf("Encountered error of type: %T", err)
```

Comment:
The `title` & `detail` fields of a JSON API error object are the most digestible and immediately obvious fields, as they have the capacity to fully describe what has taken place with an error in a completely reproducible manner. None of the other fields offer the same feature.

Keeping the code as-is also allows for developers to use the `MarshalErrors` func without having to implement any of the interfaces here (which is what you seemed to be arguing for in your above — larger — review comment.

I would like to emphasize: **the functionality here is what allows this system to be used with any error type. Without the need to implement any of the interfaces defined in this PR.**